src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,433 @@
+
+/*
+ * Copyright (C) cuiweixie
+ */
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_util.h"
+#include "ngx_http_lua_semaphore.h"
+#include "ngx_http_lua_contentby.h"
+
+int ngx_http_lua_ffi_set_semaphore_threshold(ngx_int_t threshold);
+static ngx_http_lua_semaphore_t *ngx_http_lua_alloc_sem(ngx_log_t *log);
+void ngx_http_lua_cleanup_sem(void *data);
+static void ngx_http_lua_free_sem(ngx_http_lua_semaphore_t *sem, ngx_log_t *log);
+static ngx_int_t ngx_http_lua_semaphore_resume(ngx_http_request_t *r);
+int ngx_http_lua_ffi_sem_new(ngx_http_lua_semaphore_t **psem, int n, char **errstr);
+int ngx_http_lua_ffi_sem_post(ngx_http_lua_semaphore_t *sem, int n, char **errstr);
+int ngx_http_lua_ffi_sem_wait(ngx_http_request_t *r, ngx_http_lua_semaphore_t *sem,
+                              int time, u_char *errstr, size_t *errlen);
+static void ngx_http_lua_ffi_semaphore_cleanup(void *data);
+void ngx_http_lua_ffi_semaphore_handler(ngx_event_t *ev);
+static void ngx_http_lua_sem_timeout_handler(ngx_event_t *ev);
+void ngx_http_lua_ffi_sem_gc(ngx_http_lua_semaphore_t *sem);
+
+static ngx_http_lua_semaphore_t *ngx_http_lua_sem_free_list = NULL;
+static ngx_int_t ngx_http_lua_semaphore_threshold = 0;
+static ngx_int_t ngx_http_lua_sem_free_count = 0;
+
+
+int
+ngx_http_lua_ffi_set_semaphore_threshold(ngx_int_t threshold)
+{
+    ngx_http_lua_semaphore_threshold = threshold;
+    return (int)ngx_http_lua_semaphore_threshold;
+}
+
+
+int
+ngx_http_lua_ffi_get_semaphore_threshold()
+{
+    return (int)ngx_http_lua_semaphore_threshold;
+}
+
+
+static ngx_http_lua_semaphore_t*
+ngx_http_lua_alloc_sem(ngx_log_t *log)
+{
+    ngx_http_lua_semaphore_t *sem;
+
+    if (ngx_http_lua_sem_free_list) {
+         sem = ngx_http_lua_sem_free_list;
+         ngx_http_lua_sem_free_list = ngx_http_lua_sem_free_list->next;
+         ngx_http_lua_sem_free_count--;
+
+    } else {
+
+        sem = ngx_alloc(sizeof(ngx_http_lua_semaphore_t), log);
```

Comment:
@cuiweixie I'm not suggesting creating page-size blocks in particular. Instead, I'm suggesting doing some heuristics to reduce `malloc` and `free` calls.

BTW, I don't think we should assume the underlying libc is always glibc (it could be something much worse) and even for glibc's allocator implementation it cannot defeat careful userland memory management at any rate since it lacks higher level information to be really smart. Creating so many small memory blocks in the system allocator has inevitable bookkeeping overhead. Since we are doing fixed-size (small) blocks here exclusively, my hunch is that it could be easier to do better with our own policies.

I suggest we should think more about it.